src/vcpkg/commands.portsdiff.cpp

```diff
@@ -25,78 +26,112 @@ namespace
                           [](const std::pair<const std::string, const SourceControlFileAndLocation*>& cache_entry) {
                               return cache_entry.second->source_control_file->to_version_spec();
                           });
     }
 
-    void check_commit_exists(const VcpkgPaths& paths, StringView git_commit_id)
+    bool check_commit_exists(DiagnosticContext& context,
+                             const Path& git_exe,
+                             const Path& builtin_ports_dir,
+                             StringView git_commit_id)
     {
-        static constexpr StringLiteral VALID_COMMIT_OUTPUT = "commit\n";
-        Checks::msg_check_exit(VCPKG_LINE_INFO,
-                               cmd_execute_and_capture_output(paths.git_cmd_builder(paths.root / ".git", paths.root)
-                                                                  .string_arg("cat-file")
-                                                                  .string_arg("-t")
-                                                                  .string_arg(git_commit_id))
-                                       .value_or_exit(VCPKG_LINE_INFO)
-                                       .output == VALID_COMMIT_OUTPUT,
-                               msgInvalidCommitId,
-                               msg::commit_sha = git_commit_id);
+        if (git_check_is_commit(context,
+                                git_exe,
+                                GitRepoLocator{GitRepoLocatorKind::CurrentDirectory, builtin_ports_dir},
+                                git_commit_id)
+                .value_or(false))
+        {
+            context.report_error(msgInvalidCommitId, msg::commit_sha = git_commit_id);
+            return false;
+        }
+
+        return true;
     }
 } // unnamed namespace
 
 namespace vcpkg
 {
-    PortsDiff find_portsdiff(const VcpkgPaths& paths,
-                             StringView git_commit_id_for_previous_snapshot,
-                             StringView git_commit_id_for_current_snapshot)
+    Optional<PortsDiff> find_portsdiff(DiagnosticContext& context,
+                                       const VcpkgPaths& paths,
+                                       StringView git_commit_id_for_previous_snapshot,
+                                       StringView git_commit_id_for_current_snapshot)
     {
-        check_commit_exists(paths, git_commit_id_for_previous_snapshot);
-        check_commit_exists(paths, git_commit_id_for_current_snapshot);
+        const auto& git_exe = paths.get_tool_exe(Tools::GIT, out_sink);
+        if (!check_commit_exists(context, git_exe, paths.root, git_commit_id_for_previous_snapshot) ||
+            !check_commit_exists(context, git_exe, paths.root, git_commit_id_for_current_snapshot))
+        {
+            return nullopt;
+        }
+
+        const auto maybe_previous =
+            read_ports_from_commit(context, paths, git_exe, "previous", git_commit_id_for_previous_snapshot);
```

Comment:
If a malformed commit results in bad output from read-tree, maybe we should attach the check_commit_exists to the read-tree failure case for all consumers?

Reply:
Not sure I understand this one; we already checked that it was a commit above?